metadata/filter.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import "strings"
+
+type MetaMap = map[string]string
+
+type metadata struct{}
+
+// Metadata exposes the full implementation
+type Metadata interface {
+	MetaGet
+	MetaCopy
+}
+
+// MetaGet exposes a subset, only Get* methods.
+type MetaGet interface {
+	GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap
+	GetLabelsWithPrefix(src MetaMap, prefix string) MetaMap
+}
+
+// MetaCopy exposes a subset, only Copy* methods.
+type MetaCopy interface {
+	CopyAnnotationsWithPrefix(src, dst MetaMap, prefix string)
+	CopyLabelsWithPrefix(src, dst MetaMap, prefix string)
+}
+
+// NewMetadata is a factory function that return a Metadata interface.
+func NewMetadata() Metadata {
+	return metadata{}
+}
+
+// NewMetaGet is a factory function that return a MetaGet interface.
+func NewMetaGet() MetaGet {
+	return metadata{}
+}
+
+// NewMetaCopy is a factory function that return a MetaCopy interface.
+func NewMetaCopy() MetaCopy {
+	return metadata{}
+}
+
+func (metadata) GetAnnotationsWithPrefix(src MetaMap, prefix string) MetaMap {
```

Comment:
These functions should have no receivers. These are utility functions and implementing them this way make its use more difficult. Interface, MetaMap, structs and `New*` functions should be removed from this file.

Reply:
This function (and the similar one for labels) should receive a `client.Object`, not the map itself. Right now, both functions are exactly the same no matter what you pass.

Reply:
> These functions should have no receivers. These are utility functions and implementing them this way make its use more difficult.

Actually, this implementation probably makes things much easier.
That is mainly because interface methods are called the same way as a normal function, without any difficulty.
However, if you mean the added overhead of defining an empty struct, using a receiver prefix, etc... then I agree there is some overhead cost associated with the implementation. But the overhead is minimal, straightforward idiomatic GO, and not too difficult to wrap one's head-around.

The main benefit of coding this way involves testing, because using an interface is probably a **_bazillion_** :open_mouth:   times easier to test, isolate, and mock. That way these functions/methods don't influence other tests in the codebase where they can be easily **mocked** away. Also, we probably do not want to retest these functions multiple times, for example during any unit test of functions that utilize these methods, because all the unit tests are done in `metadata/filter_test.go`.

Reply:
I am also confused as to why we have the interfaces and structs and all this extra stuff in here now. Before (I had actually approved it but missed a bug I guess) there was much less new code, I found it very easy to follow, and it only had I think two new functions. Why do we need all this extra stuff now if we didn't before?

Reply:
> I don't think it makes things easier.

**¯\\_(ツ)_/¯**

But it really does make thing easier, testing is like a **_bazillion_** times easier this way. 
I think what you probably mean is "it makes things uglier", and I would agree.

My approach is to make code easily testable, but I am trying my best to make it look less ugly. Take a look at the most recent refactor, and let me know what you think...

I've incorporated more of your suggestions.
The receiver methods are now setup to receive the `Release` & `PipelineRun` objects more directly.

Reply:
No, I meant what I said. Code should be easier to use specially with utility functions. Being easier to test comes after. I still would like to see the changes I suggested applied.

Reply:
Also, please, client.Object. Doesn't make any sense to use interfaces and templating on a utility file (which I don't agree with) and then pass specific types.

Reply:
See the most recent refactor (v).

- Method signatures uses the `Object` instead of maps.
- Should be easy to use: 
  - The `Copy*` method can do everything in one shot, or 
  - Using `Get*` + `Add*` methods together achieve the same.

PTAL